datafusion/sql/src/parser.rs

```diff
@@ -523,9 +523,7 @@ impl<'a> DFParser<'a> {
                 Ok(n) => Ok(Value::Number(n, l)),
                 // The tokenizer should have ensured `n` is an integer
                 // so this should not be possible
-                Err(e) => parser_err!(format!(
-                    "Unexpected error: could not parse '{n}' as number: {e}"
-                )),
+                Err(e) => match e {},
```

Comment:
I don't understand this change -- doesn't `match e{}` just ignore `e`  where as before this would have returned an error?

Reply:
`e` is `Infallible`, so it can't be constructed. Since it can't be constructed, this code path is unreachable, and that's what this proves. a match with no arms has a return type of `!` since it can't ever be executed, so it can be implicitly converted to any type, and can prove that a `Result` cannot be its `Err` variant.

Reply:
I think we can avoid this error handling entirely by not trying to parse a String in the first place. I'll make a follow on PR